consult-org.el

```diff
@@ -0,0 +1,167 @@
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+;;; Commentary:
+
+;; Provides a completing-read interface for some Org mode
+;; actions. This is an extra package, to allow lazy loading of Org.
+
+;;; Code:
+
+(require 'consult)
+(require 'org-agenda)
+(require 'org-clock)
+
+(defvar consult-org--history nil)
+
+(defun consult-org--narrow ()
+  "Narrowing configuration for `consult-org' commands."
+  (let ((todo-keywords (seq-filter
+                        (lambda (it) (<= ?a (car it) ?z))
+                        (mapcar (lambda (s)
+                                  (pcase-let ((`(,a ,b) (split-string s "(")))
+                                    (cons (downcase (string-to-char (or b a))) a)))
+                                (mapcan 'cdr org-todo-keywords)))))
+    (cons (lambda (cand)
+            (cond ((<= ?1 consult--narrow ?9)
+                   (<= (get-text-property 0 'consult-org--level cand)
+                       (- consult--narrow ?0)))
+                  ((<= ?A consult--narrow ?Z)
+                   (eq (get-text-property 0 'consult-org--priority cand)
+                       consult--narrow))
+                  ((when-let ((todo (alist-get consult--narrow todo-keywords)))
+                     (string-equal (get-text-property 0 'consult-org--todo cand)
+                                   todo)))))
+          (append (mapcar (lambda (c) (cons c (format "Level %c" c)))
+                          (number-sequence ?1 ?9))
+                  (mapcar (lambda (c) (cons c (format "Priority %c" c)))
+                          (number-sequence (max ?A org-highest-priority)
+                                           (min ?Z org-lowest-priority)))
+                  todo-keywords))))
+
+(defun consult-org--entries (match scope &rest skip)
+  "Return a list of consult locations from Org entries.
+
+MATCH, SCOPE and SKIP are as in `org-map-entries'."
+  (let (opoint line buffer)
+    (apply
+     #'org-map-entries
+     (lambda ()
+       (unless (eq buffer (current-buffer))
+         (setq opoint (point-min))
+         (setq line 1)
+         (setq buffer (current-buffer))
+         (setq org-outline-path-cache nil))
+       (setq line (+ line (consult--count-lines (prog1 (point)
+                                                  (goto-char opoint)))))
+       (setq opoint (point))
+       (pcase-let ((`(_ ,level ,todo ,prio) (org-heading-components)))
+         (consult--location-candidate
+          (org-format-outline-path (org-get-outline-path t t))
+          (point-marker)
+          line
+          'consult-org--level level
+          'consult-org--todo todo
+          'consult-org--priority prio)))
```

Comment:
Here only attach this single property which encodes the narrow keys.

Reply:
Most of the time, the user will not use narrowing. Why would we spend time here to precompute a value that makes it easier to narrow? It also seems nicer to keep this metadata transparent, in case the user wants to redefine their narrow criteria or if https://github.com/oantolin/orderless/issues/30 comes into being.

Or did you say single property to mean something like `'consult-org--narrow` (list level todo prio)`?

Note also that the collection of TODO keywords is customizable, and would have to been computed here again if we wanted to encode it.

Reply:
> It also seems nicer to keep this metadata transparent, in case the user wants to redefine their narrow criteria or if oantolin/orderless#30 comes into being.

Yes, then we can reconsider to add special properties with public (non double dashed) names. I mean one can always find arguments either way. I think I prefer it the other way since it is more consistent with the rest of the code.

> Or did you say single property to mean something like 'consult-org--narrow (list level todo prio)?

I meant to attach a property to the candidates which is a string of the narrowing keys, e.g. "0tA". Then the narrow function only has to check using seq-position if consult--narrow is an element of the keys.